plugins/janus_audiobridge.c

```diff
@@ -8430,11 +8166,135 @@ static void *janus_audiobridge_participant_thread(void *data) {
 	outpkt->silence = FALSE;
 	uint8_t *payload = (uint8_t *)outpkt->data;
 
+	JitterBufferPacket jbp = {0};
+	janus_audiobridge_buffer_packet *bpkt = NULL;
+	janus_audiobridge_rtp_relay_packet *pkt = NULL;
 	janus_audiobridge_rtp_relay_packet *mixedpkt = NULL;
+	janus_rtp_header *rtp = NULL;
+	gint64 now = janus_get_monotonic_time(), before = now;
+	gboolean first = TRUE, use_fec = FALSE;
+	int ret = 0;
 
-	/* Start working: check the outgoing queue for packets, then encode and send them */
+	/* Start working: check both the incoming queue (to decode and queue) and the outgoing one (to encode and send) */
 	while(!g_atomic_int_get(&stopping) && g_atomic_int_get(&session->destroyed) == 0) {
-		mixedpkt = g_async_queue_timeout_pop(participant->outbuf, 100000);
+		/* Start with packets to decode and queue for the mixer */
+		now = janus_get_monotonic_time();
+		janus_mutex_lock(&participant->qmutex);
+		gboolean locked = TRUE;
+		/* Start by reading packets to decode from the jitter buffer on a clock */
+		if(now - before >= 18000) {
+			before += 20000;
+			if(participant->jitter) {
+				ret = jitter_buffer_get(participant->jitter, &jbp, participant->codec == JANUS_AUDIOCODEC_OPUS ? 960 : 160, NULL);
+				jitter_buffer_tick(participant->jitter);
+				if(ret == JITTER_BUFFER_OK) {
+					bpkt = (janus_audiobridge_buffer_packet *)jbp.data;
+					janus_mutex_unlock(&participant->qmutex);
+					first = FALSE;
+					locked = FALSE;
+					rtp = (janus_rtp_header *)bpkt->buffer;
+					/* If this is Opus, check if there's a packet gap we should fix with FEC */
+					use_fec = FALSE;
+					if(!first && participant->codec == JANUS_AUDIOCODEC_OPUS && participant->fec) {
+						if(ntohs(rtp->seq_number) != participant->expected_seq) {
+							/* Lost a packet here? Use FEC to recover */
+							use_fec = TRUE;
```

Comment:
It's supposed to be implicit, in the sense that this check is done on an out of order packet, so the packet that follows a loss (we detect a gap). In that case, even if there was a loss higher than 1, we only inject a single FEC decoded packet anyway, since previous losses if any are ignored.